pkg/client/group_mgmt_client.go

```diff
@@ -61,46 +61,63 @@ type Argument struct {
 	JobId string `json:"job_id,omitempty"`
 }
 
-// NewClient instantiates a new client to communicate with the Nimble group
-func NewClient(ipAddress, username, password, apiVersion string, waitOnJobs bool) (*GroupMgmtClient, error) {
-	// Create GroupMgmt Client
-	client := resty.New()
-	client.SetTLSClientConfig(&tls.Config{
+func newGroupMgmtClient(ipAddress, username, password, apiVersion string, waitOnJobs bool) *GroupMgmtClient {
+	// Get new resty()
+	restyClient := resty.New()
+	restyClient.SetTLSClientConfig(&tls.Config{
 		InsecureSkipVerify: true,
 	})
+
+	// Create GroupMgmt Client
 	groupMgmtClient := &GroupMgmtClient{
 		URL:       fmt.Sprintf("https://%s:5392/%s", ipAddress, apiVersion),
-		Client:    client,
+		Client:    restyClient,
 		WaitOnJob: waitOnJobs,
 		Username:  username,
 		Password:  password,
 	}
+	return groupMgmtClient
+}
 
+// NewClient instantiates a new client to communicate with the Nimble group
+func NewClient(ipAddress, username, password, apiVersion string, waitOnJobs bool) (*GroupMgmtClient, error) {
+	// Get resty client
+	groupMgmtClient := newGroupMgmtClient(ipAddress, username, password, apiVersion, waitOnJobs)
 	// Get session token
 	sessionToken, err := groupMgmtClient.login(username, password)
 	if err != nil {
 		return nil, err
 	}
 	groupMgmtClient.SessionToken = sessionToken
 
+	// Login retry counter
+	loginRetry := 0
 	// Add retryCondition
 	groupMgmtClient.Client.
-		SetRetryCount(maxOpsRetries).
 		AddRetryCondition(
 			func(resp *resty.Response, err error) bool {
+
 				if err == nil && resp.StatusCode() == 401 {
 					// login again and refreh the session token
-					sessionToken, err = groupMgmtClient.login(username, password)
-					if err != nil {
+					loginRetry++
+					if loginRetry == maxLoginRetries {
+						// exhausted the login attempt. backoff now .
 						return false
 					}
+					newGroupMgmtClient := newGroupMgmtClient(ipAddress, username, password, apiVersion, waitOnJobs)
+					sessionToken, err = newGroupMgmtClient.login(username, password)
+					if err != nil {
+						// returning true would retry the login again with new instance of group mgmt client
```

Comment:
Why is another retry desirable?

Reply:
We want to retry until "if loginRetry == maxLoginRetries", so even if we get err via login attempt we would still retry. 
So retry backoff only when it hit loginRetry == maxLoginRetries condition.

Reply:
do you know we get same error code in both the  cases i. e expired token and invalid credentials

Reply:
Is status code 401 returned in both cases?  Yes, it must be.  The expired token case already depends on a 401 status.  The invalid credentials case would not have encountered retry recursion if it did not return status code 401.